src/TransitionGroup.js

```diff
@@ -82,126 +87,75 @@ const defaultProps = {
 
     // Initial children should all be entering, dependent on appear
     this.state = {
-      children: getChildMapping(props.children, child => {
-        return cloneElement(child, {
-          onExited: this.handleExited.bind(this, child),
-          in: true,
-          appear: this.getProp(child, 'appear'),
-          enter: this.getProp(child, 'enter'),
-          exit: this.getProp(child, 'exit'),
-        })
-      }),
-     };
+      handleExited,
+      firstRender: true,
+    }
   }
 
   getChildContext() {
     return {
-       transitionGroup: { isMounting: !this.appeared }
+      transitionGroup: { isMounting: !this.appeared },
     }
   }
-  // use child config unless explictly set by the Group
-  getProp(child, prop, props = this.props) {
-    return props[prop] != null ?
-      props[prop] :
-      child.props[prop];
-  }
 
   componentDidMount() {
-    this.appeared = true;
+    this.appeared = true
   }
 
-  componentWillReceiveProps(nextProps) {
-    let prevChildMapping = this.state.children;
-    let nextChildMapping = getChildMapping(nextProps.children);
-
-    let children = mergeChildMappings(prevChildMapping, nextChildMapping);
-
-    Object.keys(children).forEach((key) => {
-      let child = children[key]
-
-      if (!isValidElement(child)) return;
-
-      const hasPrev = key in prevChildMapping;
-      const hasNext = key in nextChildMapping;
-
-      const prevChild = prevChildMapping[key];
-      const isLeaving = isValidElement(prevChild) && !prevChild.props.in;
-
-      // item is new (entering)
-      if (hasNext && (!hasPrev || isLeaving)) {
-        // console.log('entering', key)
-        children[key] = cloneElement(child, {
-          onExited: this.handleExited.bind(this, child),
-          in: true,
-          exit: this.getProp(child, 'exit', nextProps),
-          enter: this.getProp(child, 'enter', nextProps),
-        });
-      }
-      // item is old (exiting)
-      else if (!hasNext && hasPrev && !isLeaving) {
-        // console.log('leaving', key)
-        children[key] = cloneElement(child, { in: false });
-      }
-      // item hasn't changed transition states
-      // copy over the last transition props;
-      else if (hasNext && hasPrev && isValidElement(prevChild)) {
-        // console.log('unchanged', key)
-        children[key] = cloneElement(child, {
-          onExited: this.handleExited.bind(this, child),
-          in: prevChild.props.in,
-          exit: this.getProp(child, 'exit', nextProps),
-          enter: this.getProp(child, 'enter', nextProps),
-        });
-      }
-    })
-
-    this.setState({ children });
+  static getDerivedStateFromProps(
+    nextProps,
+    { children: prevChildMapping, handleExited, firstRender }
+  ) {
+    return {
+      children: firstRender
+        ? getInitialChildMapping(nextProps, handleExited)
+        : getNextChildMapping(nextProps, prevChildMapping, handleExited),
+      firstRender: false,
```

Comment:
`getDerivedStateFromProps` is called also for the first render, so I'm not sure if you have wanted to always return `firstRender: false` here

Reply:
The idea is that `firstRender` is intialized to `true`, so if we always set it to false in in gDSFP it will only be true for the first render cycle, does that not track?

Reply:
Oh, I haven't noticed that `firstRender` is not actually used anywhere else than here, but still according to [this](https://twitter.com/dan_abramov/status/981712092611989509) `getDerivedStateFromProps` might be restarted by React, so I believe if that happens by any change (for reasons I dont know 😉) you will end up with already updated value here and unfortunately you will use `getNextChildMapping` instead `getInitialChildMapping`

Reply:
No idea, but I suspect if this hook can be called multiple times before initial render it's very likely that it gets called with updated data from previous calls and not with initial ones - it would be a little bit weird to keep both initial state & calculated state in memory.

Reply:
Ok, then maybe it's worth storing `lastChildren` as state field, call `getInitialChildMapping` in constructor, and `getNextChildMapping` when `nextProps.children !== prevState.lastChildren`